File: packages/web/src/api/post.api.ts
import { createApi } from '@reduxjs/toolkit/query/react'
import qs from 'qs'

import baseQueryWithReauth from './baseQueryWithReauth'

import { IPost } from '~types/post/post.type'

export const api = createApi({
  reducerPath: 'postApi',
  baseQuery: baseQueryWithReauth,
  endpoints: builder => ({
    getPostById: builder.query<IPost, { id: string | undefined }>({
      query: ({ id }) => ({
        url: `posts/${id}`,
        method: 'GET',
      }),
    }),
    getPosts: builder.query<IPost[], { page: number; limit: number; q?: string; tags?: string[]; sort?: string }>({
      query: ({ page, limit, q, tags, sort }) => {
        const queryString = qs.stringify({ page, limit, q, tags, sort }, { arrayFormat: 'brackets' })

        return {
          url: `posts/search?${queryString}`,
          method: 'GET',
        }
      },
    }),
    getPopularTags: builder.query<string[], void>({
      query: () => ({
        url: `posts/get-popular-tags?limit=5`,
        method: 'GET',
      }),
    }),
    createPost: builder.mutation<IPost, FormData>({
      query: body => ({
        url: 'posts',
        method: 'POST',
        body,
      }),
    }),
    dislike: builder.mutation<IPost, { postId: string; userId: string | undefined }>({
      query: ({ postId, userId }) => ({
        url: `posts/${postId}/dislike`,
        method: 'POST',
        body: { userId },
      }),
    }),
    like: builder.mutation<IPost, { postId: string; userId: string | undefined }>({
      query: ({ postId, userId }) => ({
        url: `posts/${postId}/like`,
        method: 'POST',
        body: { userId },
      }),
    }),
    editPost: builder.mutation<IPost, { postId: string | undefined; body: FormData }>({
      query: ({ postId, body }) => ({
        url: `posts/${postId}`,
        method: 'PATCH',
        body,
      }),
    }),
    deletePost: builder.mutation<void, { id: string | undefined }>({
      query: ({ id }) => ({
        url: `posts/${id}`,
        method: 'DELETE',
      }),
    }),
  }),
})

export const {
  useGetPostsQuery,
  useLazyGetPostsQuery,
  useCreatePostMutation,
  useGetPostByIdQuery,
  useDislikeMutation,
  useLikeMutation,
  useEditPostMutation,
  useDeletePostMutation,
  useGetPopularTagsQuery,
} = api

File: packages/mobile/app/src/main/java/com/devhub/devhubapp/api/PostAPI.kt
package com.devhub.devhubapp.api

import com.devhub.devhubapp.dataClasses.Post
import com.devhub.devhubapp.dataClasses.UserIdRequest
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Call
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.Multipart
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Part
import retrofit2.http.Path
import retrofit2.http.Query

interface PostAPI {
    @GET("posts")
    suspend fun getPosts(
        @Query("limit") limit: Int = 10,
        @Query("page") page: Int = 1
    ): List<Post>

    @Multipart
    @POST("posts")
    fun createPost(
        @Part("userId") userId: RequestBody,
        @Part("title") title: RequestBody,
        @Part("content") content: RequestBody,
        @Part("tags") tags: RequestBody,
        @Part headerImage: MultipartBody.Part?
    ): Call<Post>

    @Multipart
    @PATCH("posts/{id}")
    fun updatePost(
        @Path("id") postId: String,
        @Part("title") title: RequestBody,
        @Part("content") content: RequestBody,
        @Part("tags") tags: RequestBody,
        @Part headerImage: MultipartBody.Part?
    ): Call<Post>

    @DELETE("posts/{id}")
    fun deletePost(
        @Path("id") postId: String
    ): Call<Void>

    @GET("posts/{id}")
    fun getPostById(
        @Path("id") postId: String
    ): Call<Post>

    @POST("posts/{id}/like")
    fun likePost(
        @Path("id") postId: String,
        @Body requestBody: UserIdRequest
    ): Call<Post>

    @POST("posts/{id}/dislike")
    fun dislikePost(
        @Path("id") postId: String,
        @Body requestBody: UserIdRequest
    ): Call<Post>

    @GET("posts/search")
    suspend fun searchPosts(
        @Query("q") query: String?,
        @Query("tags[]") tags: List<String>?,
        @Query("sort") sort: String?,
        @Query("page") page: Int = 1,
        @Query("limit") limit: Int = 10
    ): List<Post>

    @GET("posts/get-popular-tags")
    suspend fun getTopTags(
        @Query("limit") limit: Int = 10
    ): List<String>
}